src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -252,18 +259,30 @@ protected Exception TryDeserialize(byte[] payload, out TRead msg)
         /// </summary>
         protected void HandleSendFinished(bool success)
         {
+            bool delayCompletion = false;
             TaskCompletionSource<object> origTcs = null;
             lock (myLock)
             {
                 origTcs = streamingWriteTcs;
                 streamingWriteTcs = null;
 
+                if (!success && !finished && IsClient)
+                {
+                    // We should be setting this only once per call, following writes will be short circuited.
+                    GrpcPreconditions.CheckState (delayedStreamingWriteTcs == null);
+                    delayedStreamingWriteTcs = origTcs;
+                    delayCompletion = true;
+                }
+
                 ReleaseResourcesIfPossible();
             }
 
             if (!success)
             {
-                origTcs.SetException(new InvalidOperationException("Send failed"));
+                if (!delayCompletion)
+                {
+                    origTcs.SetException(IsClient ? GetRpcExceptionClientOnly() : new IOException("Error sending from server."));
```

Comment:
Yes, you understand that correctly, but you are right that the code was a bit cryptic. I tried to improved that by adding a precondition and some comments.